File: Leetcode/199/199.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int height = -1;
    vector<vector<int> > vec;
    void dfs(TreeNode* root, int currHeight) {
        if (root) {
            dfs(root->left, currHeight + 1);
            dfs(root->right, currHeight + 1);
        } else {
            height = max(height, currHeight);
        }
    }
    void bfs(TreeNode* root) {
        if (root) {
            queue<pair<TreeNode*, int> > q;
            q.push({root, 0});
            while(!q.empty()) {
                pair<TreeNode*, int> cur = q.front();
                TreeNode* curNode = cur.first;
                int curHeight = cur.second;
                vec.at(curHeight).emplace_back(curNode->val);
                q.pop();
                if (curNode->left) {
                    q.push({curNode->left, curHeight + 1});
                }
                if (curNode->right) {
                    q.push({curNode->right, curHeight + 1});
                }
            }
        }
    }
    vector<int> rightSideView(TreeNode* root) {
        dfs(root, 0);
        vec.resize(height);
        bfs(root);
        vector<int> result(height);
        for (int i = 0; i < height; ++i) {
            result.at(i) = vec.at(i).back();
        }
        return result;
    }
};

File: Leetcode/199/199.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

func getHeight(node *TreeNode) int {
	if node == nil {
		return 0
	}
	return 1 + max(getHeight(node.Left), getHeight(node.Right))
}

type Pair struct {
	Node *TreeNode
	Level int
}

func rightSideView(root *TreeNode) []int {
	var result []int
	if root == nil {
		return result
	}
	// get height
	height := getHeight(root)
	levels := make([][]int, height)
	// bfs
	var queue []Pair
	queue = append(queue, Pair{
		Node:   root,
		Level: 0,
	})
	for len(queue) > 0 {
		front := queue[0]
		queue = queue[1:]
		levels[front.Level] = append(levels[front.Level], front.Node.Val)
		if front.Node.Left != nil {
			queue = append(queue, Pair{
				Node: front.Node.Left,
				Level: front.Level+1,
			})
		}
		if front.Node.Right != nil {
			queue = append(queue, Pair{
				Node: front.Node.Right,
				Level: front.Level+1,
			})
		}
	}
	for i := 0; i < len(levels); i++ {
		result = append(result, levels[i][len(levels[i])-1])
	}
	return result
}